instance/factories.py

```diff
@@ -0,0 +1,79 @@
+# Copyright (C) 2015-2016 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app - Factory functions for creating instances
+"""
+
+# Imports #####################################################################
+
+from django.conf import settings
+from django.template import loader
+
+from instance.models.openedx_instance import OpenEdXInstance
+
+
+# Functions ###################################################################
+
+def instance_factory(**kwargs):
+    """
+    Factory function for creating instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    When called without any arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *sandbox* instances.
+
+    To create an instance with default settings that are suitable for production,
+    use `production_instance_factory`.
+    """
+    return OpenEdXInstance.objects.create(**kwargs)
+
+
+def production_instance_factory(**kwargs):
+    """
+    Factory function for creating production instances.
+
+    Returns a newly created OpenEdXInstance.
+
+    Callers can use keyword arguments to pass in non-default values
+    for any field that is defined on the OpenEdXInstance model.
+
+    When called without any arguments, the instance that is returned
+    will have its fields set to default values that are appropriate
+    for *production* instances.
+
+    To create an instance with default settings that are suitable for sandboxes,
+    use `instance_factory`.
+    """
+    production_instance = OpenEdXInstance.objects.create(**kwargs)
+    if "use_ephemeral_databases" not in kwargs:
+        production_instance.use_ephemeral_databases = False
+    if "configuration_version" not in kwargs:
+        production_instance.configuration_version = settings.LATEST_OPENEDX_RELEASE
+    if "openedx_release" not in kwargs:
+        production_instance.openedx_release = settings.LATEST_OPENEDX_RELEASE
+    if "configuration_extra_settings" not in kwargs:
+        template = loader.get_template('instance/ansible/prod-vars.yml')
+        production_instance.configuration_extra_settings = template.render({})
```

Comment:
@antoviaque Why would the function grow over time? If we are going to eliminate the differences between sandboxes and production instances, we won't need this function anymore (or at least be able to reduce the number of settings that it customizes).

> I'd rather add a way to make the defaults depend on class attributes that can be overridden in the proxy model, for example.

What do you mean by "proxy model" here?

---

I have to admit that I am getting confused about what we are after here. Based on my current understanding, it is already possible to create instances with settings that are suitable for production; it currently just requires a few manual steps. So it seems like the goal is to be able to make it easier to create production instances (as I mentioned in my most recent [comment](https://tasks.opencraft.com/browse/OC-1554?focusedCommentId=22242&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-22242) on OC-1554); and one way of achieving that is via factory functions (such as the ones I implemented in https://github.com/open-craft/opencraft/pull/71/commits/5eb1d3227bcb0d544f8df6f4415335d9e5601018). Once we unify the settings for sandbox and production instances (which is supposed to happen when implementing continuous deployment), we can just remove the factory functions. This should be easier than having to modify the class hierarchy again at that point (which would happen if we introduced a separate class for production instances now). Does that make sense, or am I off track here?

CC @bradenmacdonald